### PR TITLE
prevent compilation for each language if error

### PR DESCRIPTION
**Pull request by OAyomide**
update lang::compile to return an Option and handle arm matching to optionally run rest of the compilation (including terraform steps) if the compilation was only successful.

i've not been able to properly test (i am not sure how to setup testing to make sure it really works) so this is somewhat a draft

### CODE REVIEW

**Comment by OAyomide** [comment]
@dotxlem tips/reviews appreciated here.

**Comment by dotxlem** [comment]
Hey @OAyomide! I don't have a very formal testing process. Typically to test changes like this I build the CLI locally, and then run the command(s) in a test project that would exercise the changes. So here I'd create a project and add a couple of functions to it, and then intentionally break one and run `cast` to make sure the process stops as intended.

There's probably a way to automate this kind of thing with a test framework, but I haven't gotten around to putting much thought into it yet 🙃

File: cli/src/commands/cast/mod.rs [hunk 1]
```diff
@@ -74,46 +74,56 @@ pub fn command(matches: Option<&ArgMatches>) {
...
+                    {
+                        let ctx = Rc::new(
+                            Context::from_project(project.clone(), asml_manifest.clone())
+                                .expect("could not make context from manifest"),
+                        );
+                        let artifacts = ctx
+                            .cast(ctx.clone(), None)
+                            .expect("could not cast assemblylift context");
+
+                        // todo: serialize to string only if the file is not out of date
```
Comment (dotxlem) [inline cli/src/commands/cast/mod.rs:129]: I think we can use the sha2 crate for this to compare the sha256 of the existing file contents and the new artifact

https://github.com/RustCrypto/hashes

Comment (OAyomide) [reply cli/src/commands/cast/mod.rs:129]: heyy. i didnt realize you already replied/reviewed. plus ive been mad busy jajajaja. i'll take a look, update and revert here again.

**Comment by OAyomide** [comment]
@dotxlem i think this is ready (for the most part, at least). next is the hash crate to use but 👀 on this please.


Screenshot 2023-04-24 at 23 07 42

**Comment by dotxlem** [comment]
Thanks @OAyomide! I’ll look at this over the weekend

**Review by dotxlem — CHANGES_REQUESTED** [review]
This is great so far @OAyomide! There's just one thing I can see that's got to do with some hacky code I added a while ago xD

At rust.rs#L47 there is:
```rust
if build_log.contains("error") {
    std::process::exit(-1);
}
```

which was a low-effort way to detect a build error in Rust specifically that I never got around to improving. It has several problems, one being that the word "error" can occur in the output regardless of a real error, and two being that it just exists the process.

So now that you've refactored this to return an Error, I think here we can return an Error instead of exiting, by checking the Cargo build exit code; something like `cargo_build.status.code().unwrap() != 0` to treat any non-zero exit code as an error.

**Comment by OAyomide** [comment]
hallo @dotxlem  been a few months but i have some additional commits that should take care of some things. i am looking for reviews before i continue with it

**Review by dotxlem — APPROVED** [review]
Hey @OAyomide, sorry it's taken me so long to get back to this. I haven't been able to work on AssemblyLift as much lately, but I'm getting back into it again slowly! I'm going to merge this now as I'm also looking to merge #111 & the beta branch into mainline.

If there are any issues here I'll handle them after the merge, but I think this all looks good! Thanks for the contribution!

**Pull request merged**